File: haskell-src/Bench.hs
import Criterion.Main
import qualified D1Lib
import qualified D2Lib
import qualified D3Lib
import qualified D4Lib
import qualified D5Lib
import qualified D6Lib
import qualified D7Lib
import qualified System.IO

type BenchData = (String, String -> Int, String -> Int)

benchData :: [BenchData]
benchData =
    [ ("d1", D1Lib.part1, D1Lib.part2)
    , ("d2", D2Lib.part1, D2Lib.part2)
    , ("d3", D3Lib.part1, D3Lib.part2)
    , ("d4", D4Lib.part1, D4Lib.part2)
    -- , ("d5", D5Lib.part1, D5Lib.part2)
    , ("d6", D6Lib.part1, D6Lib.part2)
    , ("d7", D7Lib.part1, D7Lib.part2)
    ]

buildBench :: BenchData -> Benchmark
buildBench (day, p1, p2) = env setupEnv setupGroup
  where
    setupGroup contents = bgroup day [bench "p1" $ whnf p1 contents, bench "p2" $ whnf p2 contents]
    setupEnv = readFile ("input/" ++ day ++ ".txt")

main = defaultMain $ map buildBench benchData

File: rust-src/bench.rs
mod d1lib;
mod d2lib;
mod d3lib;
mod d4lib;

use criterion::{black_box, criterion_group, criterion_main, Criterion};
use std::fs::read_to_string;

const BENCH_DATA: &[(&str, fn(&str) -> usize, fn(&str) -> usize)] = &[
    ("d1", d1lib::part1, d1lib::part2),
    ("d2", d2lib::part1, d2lib::part2),
    ("d3", d3lib::part1, d3lib::part2),
    ("d4", d4lib::part1, d4lib::part2),
];

pub fn run_benchmarks(c: &mut Criterion) {
    for (day, p1, p2) in BENCH_DATA.iter().copied() {
        let filename = format!("input/{day}.txt");
        let file_contents = read_to_string(filename).unwrap();
        c.bench_function(&format!("{day}p1"), |b| {
            b.iter(|| black_box(p1(&file_contents)))
        });
        c.bench_function(&format!("{day}p2"), |b| {
            b.iter(|| black_box(p2(&file_contents)))
        });
    }
}

criterion_group!(benches, run_benchmarks);
criterion_main!(benches);
